aptos-move/framework/supra-framework/sources/automation_registry.move

```diff
@@ -0,0 +1,152 @@
+        /// Resource account signature capability
+        registry_fee_address_signer_cap: SignerCapability,
+        /// A collection of automation task entries that are active state.
+        tasks: EnumerableMap<u64, AutomationTaskMetaData>,
+    }
+
+    #[event]
+    /// `AutomationTaskMetaData` represents a single automation task item, containing metadata.
+    struct AutomationTaskMetaData has copy, store, drop {
+        /// Automation task index in registry
+        id: u64,
+        /// The address of the task owner.
+        owner: address,
+        /// The function signature associated with the registry entry.
+        payload_tx: vector<u8>,
+        /// Expiry of the task, represented in a timestamp in second.
+        expiry_time: u64,
+        /// The transaction hash of the request transaction.
+        tx_hash: vector<u8>,
+        /// Max gas amount of automation task
+        max_gas_amount: u64,
+        /// Maximum gas price cap for the task
+        gas_price_cap: u64,
+        /// Registration epoch number
+        registration_epoch: u64,
+        /// Registration epoch time
+        registration_time: u64,
+        /// Flag indicating whether the task is active.
+        is_active: bool
+    }
+
+    // todo : this function should call during initialzation, but since we already done genesis in that case who can access the function
+    fun initialize(supra_framework: &signer) {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let (registry_fee_resource_signer, registry_fee_address_signer_cap) = account::create_resource_account(
+            supra_framework,
+            REGISTRY_RESOURCE_SEED
+        );
+
+        move_to(supra_framework, AutomationRegistry {
+            current_index: 0,
+            automation_gas_limit: DEFAULT_AUTOMATION_GAS_LIMIT,
+            duration_upper_limit: DEFAULT_DURATION_UPPER_LIMIT,
+            registry_fee_address: signer::address_of(&registry_fee_resource_signer),
+            registry_fee_address_signer_cap,
+            tasks: enumerable_map::new_map(),
+        })
+    }
+
+    // todo withdraw amount from resource accoun to specified address
+    fun withdraw() {}
+
+
+    public(friend) fun on_new_epoch() {
+        // todo : should perform clean up and updation of state
+    }
+
+    /// Registers a new automation task entry.
+    public entry fun register(
+        owner: &signer,
+        payload_tx: vector<u8>,
+        expiry_time: u64,
+        max_gas_amount: u64,
+        gas_price_cap: u64
+    ) acquires AutomationRegistry {
+        let registry_data = borrow_global_mut<AutomationRegistry>(@supra_framework);
+
+        // todo : well formedness check of payload_tx
+        // todo : pre-paid amount collect from the user
+        // todo : duration/expiry in seconds
+        // Expiry time does not go beyond upper cap duration set by admin/governance
+        assert!(expiry_time < registry_data.duration_upper_limit, EEXPIRY_TIME_UPPER);
```

Comment:
I assume that the client will provide the expiration time in seconds, indicating how many days or months from now they need the execution to occur.
For example, if they want the execution to be valid for only one day, they would provide 86,400 seconds as an argument.

Please let me know if we should handle it in the manner you’ve outlined.
CC: Dr. @sjoshisupra

Reply:
Based on the last discussion I remember that we decided to have exact exipration-time in seconds.
`Now` can be different in scope of different nodes, so global/absolute time is better than relative one. and this value also will be used filter out already expired automated-transactions in scope of SMR absolute time is preferable.
So I think even here we should take `block-creation-time` or `predicted_next_epoch_start_time` to check the actual duration, instead of `current-time`